Fix Stripe Billing checkout with different currencies

reykjalin opened the pull request:
Fixes #7732

## Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

* Create new `Cannot_Combine_Currencies_Exception`.
* Throw this exception when the following conditions are met:
    1. The merchant has Stripe Billing enabled;
    2. The merchant has multi-currency enabled with at least 2 available currencies;
    3. The customer has a pre-existing, active subscription with currency A;
    4. The customer tries to buy a new subscription with currency B; and
    5. The Stripe API returns an error indicating that a customer attempted to checkout with a new currency.
* Catch the exception in the subscription service and throw an error that informs the customer of why the payment failed.

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Setup

1. Checkout `develop`.
2. Make sure you have WC Subscriptions installed on your test store.
3. Make sure you have a subscription product available, a simple subscription is sufficient.
4. Enable Stripe Billing in the WooPayments settings.
5. Enable multi-currency and enable any 2 currencies (e.g. USD and EUR)


### Reproduce

1. Add a subscription product to your cart, and checkout with any card, `4242 4242 4242 4242` is sufficient.
2. The checkout should succeed.
3. Change the currency.
    * You can do this either by using the currency switching widget or by visiting any store page with `?currency=<currency_code>`, e.g. `?currency=EUR`.
4. Add the same subscription product to your cart and checkout with any card, e.g. the card that was saved after checking out in (1).
5. Read the error message:

| | Before (`develop`) | After |
|:---:|:---:|:---:|
| Shortcode | ![image](https://github.com/Automattic/woocommerce-payments/assets/13835680/11c39f72-791f-4c3b-a027-b6ed06e12f35) | ![image](https://github.com/Automattic/woocommerce-payments/assets/13835680/9a00cf86-4a83-4732-911a-6de1260db5ec) |
| Blocks | ![image](https://github.com/Automattic/woocommerce-payments/assets/13835680/465e7336-92b1-416f-903d-9a15be8bc3fc) | ![image](https://github.com/Automattic/woocommerce-payments/assets/13835680/a39550bc-8cf2-436c-8bd8-f2b34ab05649) |



-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-7.5.0#make-sure-an-actionable-error-is-reported-when-trying-to-buy-a-subscription-through-stripe-billing-with-a-different-currency-from-your-active-subscriptions
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. **Not applicable**
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
add scaffolding to reproduce issue in an e2e test

Commit:
Add working Stripe billing test

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8462` or branch name `fix/stripe-billing-checkout-with-different-currencies` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/stripe-billing-checkout-with-different-currencies) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 053cb5de7a231604ae05ba04f0e0c76056e70700
- Build time: 2024-04-03 16:46:07 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Commit:
Add Cannot_Combine_Currencies_Exception class

This exception will be thrown when a customer tries to create a Stripe
Billing subscription with a currency that doesn't match the currency on
an already active subscription through Stripe Billing.

Commit:
Throw Cannot_Combine_Currencies_Exception when using invalid currency for Stripe Billing subscription

Commit:
try to add e2e test for the failure

Commit:
fix incorrect currency type in exception

Commit:
Merge branch 'develop' into fix/stripe-billing-checkout-with-different-currencies

Commit:
Move new e2e tests to the subscriptions suite of tests

reykjalin commented:
@ricardo When I run the tests individually - either via CLI filter or by using `it.only` - they run to completion but when I run them together they always fail. Running them with `npm run test:e2e-dev` causes a segfault in Chromium.

I have no idea why. Can you see if you can run the e2e tests properly?

Commit:
add changelog entry

reykjalin commented:
Marked with `needs design` because I'd like to get design to review the language in the error message. Asked in p1711762260578259-slack-C01TNBRTMEU.

Commit:
add translators comment

ricardo commented on includes/subscriptions/class-wc-payments-subscription-service.php:
```diff
@@ -417,6 +418,15 @@ public function create_subscription( WC_Subscription $subscription ) {
+				throw new Exception(
+					sprintf(
+						// Translators: %1$s and %2$s are both currency codes, e.g. `USD` or `EUR`.
+						__('There was a problem creating your subscription. All your active subscriptions must use the same currency. You attempted to purchase a subscription in %1$s but have another active subscription using %2$s.', 'woocommerce-payments' ),
```
🤏 Nit: Missing space after opening parenthesis.

reykjalin replied:
Hmmm, I'm realizing PHPCS doesn't catch this anymore 🤔 

Not even if I revert 660a6302dc439aed8751c547952d8724647bb475 to go back to before we updated the sniffs.

Not sure why. Huh.

reykjalin replied:
Fixed in bcd7fdc3a2957726fb46cafe883bd2aec180e3dc

reykjalin replied:
Opened #8533 to track the phpcs issue

ricardo commented:
> Can you see if you can run the e2e tests properly?

@reykjalin The new test spec looks good, but I couldn't get my test environment to work with Stripe billing.

When I run `npm run test:e2e-dev tests/e2e/specs/subscriptions/shopper/shopper-subscriptions-multi-currency.spec.js`, it fails with the following error for both tests:

> There was a problem creating your subscription. Please try again or contact us for assistance.

This is probably due to some issue with my local e2e environment, because the PR tests well on my Valet WooPayments environment.

ricardo reviewed (approved):
Tests well and LGTM.

I get the following error notice when trying to pay with a different currency:

<img width="100%" alt="Stripe Billing checkout with different currencies Notice" src="https://github.com/Automattic/woocommerce-payments/assets/5509901/629ff34d-fb42-4808-a526-c9836fdf6a84">

The only thing to mention is that I also couldn't run the e2e tests as you requested. Maybe we could ask someone else from Fractal before merging? I need more time to troubleshoot why my local e2e tests env is not working.

reykjalin committed:
Merge branch 'develop' into fix/stripe-billing-checkout-with-different-currencies

Commit:
Add missing space in function call

Commit:
Remove e2e tests added

They kept crashing and it doesn't make sense to spend more time on them at
this point.

Commit:
Update error language to match direction from design

reykjalin committed:
Merge branch 'develop' into fix/stripe-billing-checkout-with-different-currencies

ricardo reviewed (approved):
LGTM 👍.
